File: leetcode2/DesignBrowserHistory/a.cpp
/* Copyright (C) 2020 by iamslash */

#include <cstdio>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

// 174ms 66.73% 55MB 80.16%
// linear traversal
// visit: O(N) O(N)
class BrowserHistory {
 private:
  vector<string> hist;
  int pos = 0;
 public:
  BrowserHistory(string url) {        
    hist.push_back(url);
    pos = 0;
  }      
  void visit(string url) {
    pos++;    
    hist.resize(pos);
    hist.push_back(url);
  }
    
  string back(int steps) {
    pos = max(0, pos - steps);
    return hist[pos];
  }
    
  string forward(int steps) {
    pos = min((int)hist.size()-1, pos + steps);
    return hist[pos];
  }
};

File: leetcode2/DesignBrowserHistory/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2020 by iamslash 

# list
# 216ms 85.81% 17MB 8.27%
class BrowserHistory:    

    def __init__(self, homepage: str):
        self.pos = 0
        self.hist = [homepage]

    def visit(self, url: str) -> None:
        self.pos += 1
        self.hist = self.hist[:self.pos]
        self.hist.append(url)

    def back(self, steps: int) -> str:
        self.pos = max(0, self.pos - steps)
        return self.hist[self.pos]

    def forward(self, steps: int) -> str:
        self.pos = min(len(self.hist)-1, self.pos + steps)
        return self.hist[self.pos]
